Add type hints and strict types to Controller

cleptric opened the pull request:
Refs: #11935 

EDIT: I slipped in one addition from the `Core` package 😇 
`InstanceConfigTrait:: getConfig `

Commit:
Add type hints and strict types to Controller

stickler-ci committed:
Fixing style errors.

Commit:
Modify stickler config

markstory commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -403,10 +406,10 @@ protected function _isLoginAction(Controller $controller)
      * @throws \Cake\Http\Exception\ForbiddenException
      */
-    protected function _unauthorized(Controller $controller)
+    protected function _unauthorized(Controller $controller): ?Response
```
Does this need to nullable if the method sometimes throws an exception?

dereuromark replied:
No, it should always be either Response, or the exception to bail early.

Commit:
Remove spaces around declare statement

markstory commented on src/Controller/Component/SecurityComponent.php:
```diff
@@ -175,7 +176,7 @@ public function blackHole(Controller $controller, $error = '', SecurityException
      */
-    protected function _throwException($exception = null)
+    protected function _throwException(?\Cake\Controller\Exception\SecurityException $exception = null): void
     {
```
Are we able to use the imported class name that isn't fully qualified?

dereuromark replied:
We should always have the FQN in the use statement IMO.

cleptric replied:
Definitely, I overlooked this one.

Commit:
Remove double space and fix return type hint

markstory commented:
You might need to use `phpcs.xml.dist` for stickler. There isn't a phpcs.xml file in the repository.

cleptric commented:
@markstory ... sorry

Commit:
Use correct phpcs.xml.dist

dereuromark commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -817,7 +821,7 @@ public function constructAuthenticate()
             }
             $className = App::className($class, 'Auth', 'Authenticate');
-            if (!class_exists($className)) {
+            if ($className === false || !class_exists($className)) {
```
Dont u need the same in line 865 too?

dereuromark commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -817,7 +821,7 @@ public function constructAuthenticate()
+            if ($className === false || !class_exists($className)) {
                 throw new Exception(sprintf('Authentication adapter "%s" was not found.', $class));
             }
             if (!method_exists($className, 'authenticate')) {
```
Do we need to keep this check in 4.x+? Maybe we can also use some interface approach in the future.

ADmad replied:
AuthComponent and related classes themselves will go away in a future release, so not worth the effort.

cleptric replied:
@ADmad Done, let me know if it's ok this way :)

Commit:
Don't use the FQC as typehint

Commit:
Add another false check

Commit:
Remove standard again and fix it on a alter date

stickler-ci committed:
Fixing style errors.

dereuromark commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -403,10 +406,10 @@ protected function _isLoginAction(Controller $controller)
      *
      * @param \Cake\Controller\Controller $controller A reference to the controller object
-     * @return \Cake\Http\Response
+     * @return \Cake\Http\Response|null
```
Must be reverted.

cleptric replied:
`Controller::redirect` could return `null` so this should be fine

dereuromark replied:
Why can it? It should either return a response or throw an exception, right?

cleptric replied:
Because of `Controller::redirect():583` 

```
if ($event->isStopped()) {
  return null;
}
```

dereuromark replied:
OK

dereuromark commented on tests/test_app/TestApp/Controller/PostsController.php:
```diff
@@ -34,7 +35,7 @@ public function initialize()
      *
      * @return void
```
should then be Response|null as usual

ADmad commented on src/Controller/Controller.php:
```diff
@@ -178,7 +179,7 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
      * @param \Cake\Controller\ComponentRegistry|null $components The component registry. Defaults to a new instance.
      */
-    public function __construct(ServerRequest $request = null, Response $response = null, $name = null, $eventManager = null, $components = null)
+    public function __construct(ServerRequest $request = null, Response $response = null, ?string $name = null, ?\Cake\Event\EventManager $eventManager = null, ?\Cake\Controller\ComponentRegistry $components = null)
```
Need to add `use` statements instead of using FQCNs. Also `EventManager` needs to be changed to `EventManagerInterface`.

ADmad commented on src/Controller/Controller.php:
```diff
@@ -249,7 +250,7 @@ public function initialize()
      * @return \Cake\Controller\ComponentRegistry
      */
-    public function components($components = null)
+    public function components(?\Cake\Controller\ComponentRegistry $components = null): ComponentRegistry
```
Need to add `use` statements instead of using FQCN.

Commit:
Fix annotations and type hints

dereuromark commented on tests/TestCase/Controller/ControllerTest.php:
```diff
@@ -632,7 +632,7 @@ public function testSetAction(): void
     public function testStartupProcess(): void
     {
-        $eventManager = $this->getMockBuilder('Cake\Event\EventManager')->getMock();
+        $eventManager = $this->getMockBuilder('Cake\Event\EventManagerInterface')->getMock();
```
We can use ::class usage here too for better refactoring.

Commit:
Fix more failing tests

stickler-ci committed:
Fixing style errors.

dereuromark commented on src/Controller/Component/RequestHandlerComponent.php:
```diff
@@ -121,7 +122,7 @@ public function implementedEvents()
      * @return void
      */
-    protected function _setExtension($request, $response)
+    protected function _setExtension(\Cake\Http\ServerRequest $request, \Cake\Http\Response $response): void
```
You can use my FQCN to use statement sniff if you dont want to do that manually :)

dereuromark commented on src/Controller/Component/RequestHandlerComponent.php:
```diff
@@ -238,7 +239,7 @@ public function convertXml($xml)
      *
      * @param \Cake\Event\EventInterface $event The Controller.beforeRender event.
-     * @return bool false if the render process should be aborted
+     * @return bool|void false if the render process should be aborted
```
Response|null actually

dereuromark replied:
Fairy certain that is a dead relict from 2.x times.
Just void should be fine for these.

ADmad replied:
It's not a relic. If a listener callback returns `false` the event manager stops event propagation.  `RequestHandler::beforeRender()` checks the last modified headers and if needed stops the event propagation to prevent view rendering and just return response with appropriate headers.

The `return false` could however be replaced with `$event->stopPropgation()` to achieve the same result.

dereuromark commented on src/Controller/ComponentRegistry.php:
```diff
@@ -52,9 +54,9 @@ public function __construct(Controller $controller = null)
+     * @return \Cake\Controller\Controller|null Controller instance
      */
-    public function getController()
+    public function getController(): ?Controller
```
Should this instead throw an exception if getController() cannot return sth?

dereuromark replied:
OK, we can allow this to be nullable I guess.

Commit:
Fix another FQN

Commit:
Revert added typehint

Commit:
Rework beforeRender

Commit:
Merge branch 4.x

Commit:
Fix annotation

Commit:
Fix typo

Commit:
Use correct typehint

Commit:
Fix failing tests

Commit:
Not longer possible

Commit:
Fix order

Commit:
Fix null when calling explode

cleptric commented:
`TypeError: explode() expects parameter 2 to be string, null given`

Looks like we need to add the content type to the `IntegrationTestCase`.
Or would a `list($contentType) = explode(';', $request->contentType() ?? '')`, as it was basically this before adding strict types.

markstory commented on tests/test_app/TestApp/Controller/AuthTestController.php:
```diff
@@ -41,7 +42,7 @@ public function __construct($request = null, $response = null)
     }
 
-    public function initialize()
+    public function initialize(): void
```
Is updating the various hook methods in an application going to add value or frustration? I left return typehints off of the hook methods in `Shell` and `Command` as I didn't see them adding much value to a developer's life, but they certainly are a big bump when upgrading.

dereuromark replied:
I also see your argument.
But: Using automated sniffers/fixers (existing), it is rather easy to fix up project level or plugins at once. 

The benefit of having a consistent code base is a requirement for this, though. If only some methods had typehinting the sniffer would fix too much and then you run into fatal errors as signature changes.

Also: The benefit for hooks is especially important (since they cannot be IDE completed/validated). People might try to return a response object or other weird things. So having a clear communicated API for them seems very useful.

markstory replied:
> The benefit of having a consistent code base is a requirement for this, though. If only some methods had typehinting the sniffer would fix too much and then you run into fatal errors as signature to parent is not compatible.

Sub-classes can add additional type restrictions that the parent class doesn't have in general.

> People might try to return a response object or other weird things. So having a clear communicated API for them seems very useful.

This could be a valuable point of having more types on hooks. It makes the framework's expectations of userland code clear. My concern is primarily around the initial upgrade and how bumpy that is going to be.

dereuromark replied:
Signature changes should be regex fixable etc.

cleptric committed:
Merge branch '4.x' into controller-type-hints

cleptric commented:
I fixed the conflict. Should be good to go

Commit:
Fix failing test

cleptric committed:
Merge branch '4.x' into controller-type-hints

markstory committed:
Restore change in referer() signature.

In #12320 the default value for local was changed to `true` this change
makes using referer safer.
